### PR TITLE
ISSUE-94 Keystore/set destroy method

### PR DESCRIPTION
**Pull request by LanKhuat**
a few concerns:

- Since we never check for the key validity, i don't think we will ever run into a "notDestroyed" case.
- Do we allow user to destroy all of their keys? What would be the input parameters in that case? Also if we succeed in deleting all of the keys, how would we be able to return all of the destroyed keys' ID?

### CODE REVIEW

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 1]
```diff
@@ -500,6 +500,224 @@ trait LinagoraKeystoreSetMethodContract {
...
+                     |			"create": {
+                     |        "K87": {
+                     |          "key": "$PGP_KEY_ARMORED"
+                     |        }
+                     |      }
+                     |		}, "c1"],
+                     |    ["Keystore/set", {
+                     |			"accountId": "$ACCOUNT_ID",
+                     |			"destroy": [
+                     |        "$PGP_KEY_ID"
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:519]: You can use `#K87` (client ID reference)

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 2]
```diff
@@ -500,6 +500,224 @@ trait LinagoraKeystoreSetMethodContract {
...
+                     |      ]
+                     |		}, "c2"]
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:521]: Indent looks weird.

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 3]
```diff
@@ -500,6 +500,224 @@ trait LinagoraKeystoreSetMethodContract {
...
+         |          "id": "$PGP_KEY_ID"
+         |        }
+         |		  }
+         |		}, "c1"],
+         |    ["Keystore/set", {
+         |		  "accountId": "$ACCOUNT_ID",
+         |			"destroyed": [
+         |        "$PGP_KEY_ID"
+         |      ]
+         |		}, "c2"]
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:555]: idem, review the indent

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 4]
```diff
@@ -500,6 +500,224 @@ trait LinagoraKeystoreSetMethodContract {
...
+                     |      }
+                     |		}, "c1"],
+                     |    ["Keystore/set", {
+                     |			"accountId": "$ACCOUNT_ID",
+                     |			"destroy": [
+                     |        "$PGP_KEY_ID"
+                     |      ]
+                     |		}, "c2"]
+                     |	]
+                     |}""".stripMargin
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:523]: Can we add a `Keystore/get` in the call? to ensure that the key have well been destroyed?

File: openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  def retrieveDestroyed: List[DestroySuccess] = values.flatMap(value => {
+    value match {
+      case p: DestroySuccess => Some(p)
+      case _ => None
+    }
+  })
+
+  def retrieveNotDestroyed: List[DestroyFailure] = values.flatMap(value => {
+    value match {
+      case p: DestroyFailure => Some(p)
```
Comment (chibenwa) [inline openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala:32]: p ?

File: openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      case failure: DestroyFailure => DestroyResults(destroyResults.retrieveNotDestroyed ++ List(failure))
+    }
+
+  def empty: DestroyResults = DestroyResults(List())
+}
+
+case class DestroyResults(values: List[DestroyResult]) {
+  def retrieveDestroyed: List[DestroySuccess] = values.flatMap(value => {
+    value match {
+      case p: DestroySuccess => Some(p)
```
Comment (chibenwa) [inline openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala:25]: p?

File: openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/model/KeystoreSet.scala [hunk 7]
```diff
@@ -14,7 +14,8 @@ import play.api.libs.json.JsObject
 case class KeystoreCreationId(id: Id)
 
 case class KeystoreSetRequest(accountId: AccountId,
-                              create: Option[Map[KeystoreCreationId, JsObject]]) extends WithAccountId
+                              create: Option[Map[KeystoreCreationId, JsObject]],
+                              destroy: Option[List[KeystoreCreationId]]) extends WithAccountId
```
Comment (chibenwa) [inline openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/model/KeystoreSet.scala:18]:
```suggestion
                              destroy: Option[List[KeyId]]) extends WithAccountId
```

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala (hunk 2 above)
Comment (LanKhuat) [reply openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:521]: likely a GitHub issue. @Arsnael did you find a way to fix this?

Comment (chibenwa) [reply openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:521]: Look if you use TAB, maybe?

**Review by chibenwa — COMMENTED** [review]
Looks good to me

File: openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    case value: DestroyFailure => Some(value)
+    case _ => None
+  }
+}
+
+sealed trait DestroyResult
+case class DestroySuccess(id: KeyId) extends DestroyResult
+case class DestroyFailure(id: KeyId, throwable: Throwable) extends DestroyResult
+
+class KeystoreSetDestroyPerformer @Inject()(serializer: KeystoreSerializer,
```
Comment (vttranlina) [inline openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetDestroyPerformer.scala:38]: remove `serializer: KeystoreSerializer` if it doesn't be used

**Review by chibenwa — APPROVED** [review]

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 9]
```diff
@@ -481,22 +522,247 @@ trait LinagoraKeystoreSetMethodContract {
...
+         |  "methodResponses": [
+         |    ["error", {
+         |      "type": "accountNotFound"
+         |    }, "c1"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def keystoreSetDestroyWithShouldFailWhenMissingOneCapability(): Unit = {
```
Comment (Arsnael) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:691]:
```suggestion
  def keystoreSetDestroyShouldFailWhenMissingOneCapability(): Unit = {
```

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 10]
```diff
@@ -481,22 +522,247 @@ trait LinagoraKeystoreSetMethodContract {
...
+         |  ]
+         |}""".stripMargin)
+
+    assertThat(server.getProbe(classOf[JmapGuiceKeystoreManagerProbe])
+      .retrieveKey(BOB, KeyId(PGP_KEY_ID)))
+      .isEmpty
+  }
+
+  @Test
+  def keystoreSetDestroyWithUnknownKeyIdShouldSucceed(): Unit = {
```
Comment (Arsnael) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:612]: Hmmmm.... really? Back to the core specs: https://jmap.io/spec-core.html#set

```
If an id given cannot be found, the update or destroy MUST be rejected with a notFound set error.
```
And that believe that's what we do with mailbox and email

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala [hunk 11]
```diff
@@ -481,22 +522,247 @@ trait LinagoraKeystoreSetMethodContract {
...
+         |    ["error", {
+         |      "type": "unknownMethod",
+         |      "description": "Missing capability(ies): com:linagora:params:jmap:pgp"
+         |    },"c1"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def keystoreSetDestroyWithShouldFailWhenMissingAllCapabilities(): Unit = {
```
Comment (Arsnael) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:731]:
```suggestion
  def keystoreSetDestroyShouldFailWhenMissingAllCapabilities(): Unit = {
```

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala (hunk 10 above)
Comment (chibenwa) [reply openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:612]: Well, idempotence is a nice behaviour too. What we can is that the stuff that the user wanted to delete is no longer here.

Also, read-before-write is a classic anti-pattern in Cassandra.

I'm okay being slightly off-spec regarding this (as it is a custom extension).

Comment (Arsnael) [reply openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:612]: I believe any custom extension should still respect the core specs at the end... But I understand the anti-pattern in Cassandra this generates as well.

Ok for now

**Review by Arsnael — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Pull request merged**